### PR TITLE
Lifecycle Nodes

### PR DESCRIPTION
**Pull request by supex0fan**
## What type of PR is this? (check all applicable)
Feature

## Description / Documentation
Lots of lines changed but fundamentally its a pretty simple change. Lifecycle nodes just allow other nodes to activate and deactivate it. The LifecycleNode provides several method that need to be overridden by classes that inherit it, these are on_configure, on_activate, on_deactivate, on_shutdown, on_cleanup and on_error. These correspond with states in the LifecycleNode state machine, they are called when the state is 'transitioned' into, which is initiated by the node's manager. In these state callbacks, we simply activate/deactivate publishers and subscribers, rendering the node enabled or disabled. In our case, the trackdrive_handler_node is the manager of three lifecycle nodes, a pure pursuit node, a reactive controller and a path planner. A convenience class called LifecycleServiceClient has been created to simplify the managing of LifecycleNodes. 

## To Test / Run
1. Run the sim
2. Run particle pursuit
3. Run reactive controller
4. Run ordered mid spline
5. Run trackdrive handler

The reactive controller should perform one lap, then it will be deactivated, and the trackdrive handler will activate the pure pursuit and ordered mid spline nodes. After 10 laps all nodes (except trackdrive handler) should be shutdown.

## Usability concerns or breaking changes?
Shouldn't be. We will see.

### CODE REVIEW

**Comment by b1n-ch1kn** [comment]
Unrelated to the PR content, but I like the "To test" section of the PR - would be worth adding to the template

**Comment by supex0fan** [comment]
> Unrelated to the PR content, but I like the "To test" section of the PR - would be worth adding to the template

Run instructions are nice too

**Review by supex0fan — COMMENTED** [review]
I actually intended to read a pure pursuit implementation from the launch args, but found the args very hard to use.

**Review by b1n-ch1kn — CHANGES_REQUESTED** [review]
Look good, see comments

File: src/control/controllers/controllers/node_reactive_control.py [hunk 1]
```diff
@@ -156,7 +123,41 @@ def callback(self, msg: ConeDetectionStamped):
...
+
+    def on_activate(self, state: LifecycleState) -> TransitionCallbackReturn:
+        self.get_logger().info("on_activate")
+        topic = "/lidar/cone_detection" if self.ebs_test else "/slam/local_map"
+        self.map_sub = self.create_subscription(ConeDetectionStamped, topic, self.callback, QOS_ALL)
+        return super().on_activate(state)
+
+    def on_deactivate(self, state: LifecycleState) -> TransitionCallbackReturn:
+        self.get_logger().info("on_deactivate")
+        self.destroy_subscription(self.map_sub)
```
[the remark this answers is not in the document]
Comment (supex0fan) [reply src/control/controllers/controllers/node_reactive_control.py:142]: https://github.com/ros2/demos/blob/36a4c0bd6ae0531ad3096806c41efd7062a1cb31/lifecycle_py/lifecycle_py/talker.py#L81-L82
Happens automatically in rclpy.

File: src/common/driverless_common/driverless_common/shutdown_node.py
[the remark this answers is not in the document]
Comment (supex0fan) [reply src/common/driverless_common/driverless_common/shutdown_node.py:1]: Personally I think we can get rid of it and implement its functionality into the trackdrive handler. Let me know and I'll do it.

[the remark this answers is not in the document]
Comment (b1n-ch1kn) [reply src/common/driverless_common/driverless_common/shutdown_node.py:1]: I like the sound of that tbh. So long as comms are maintained between hander and process node it should be able to confirm shutdown. 
This happened for the first time when i hit lap 10 in testing and trackdrive couldnt finish/shutdown cause it was still polling the pursuit node

image
image

File: src/control/path_follower/path_follower/node_particle_pursuit_inherit.py [hunk 2]
```diff
@@ -240,6 +239,35 @@ def publish_debug_image(self, steering_angle: float, velocity: float, rvwp: List
 
         self.debug_publisher.publish(cv_bridge.cv2_to_imgmsg(debug_img, encoding="bgr8"))
 
+    def publish_debug_image(self, steering_angle: float, velocity: float, rvwp: List[float], position: List[float]):
```
Comment (b1n-ch1kn) [inline src/control/path_follower/path_follower/node_particle_pursuit_inherit.py:242]: Just noticed this function is here twice? the only change is the publisher name, which i think is whats breaking

Comment (b1n-ch1kn) [reply src/control/path_follower/path_follower/node_particle_pursuit_inherit.py:242]: That was not the issue...

Comment (supex0fan) [reply src/control/path_follower/path_follower/node_particle_pursuit_inherit.py:242]: You talking about the issue in the other comment?

File: src/common/driverless_common/driverless_common/shutdown_node.py
[the remark this answers is not in the document]
Comment (b1n-ch1kn) [reply src/common/driverless_common/driverless_common/shutdown_node.py:1]: Looks like this is to do with debug image publishers
https://discourse.ros.org/t/about-qos-of-images/18744

File: src/control/path_follower/path_follower/node_particle_pursuit_inherit.py (hunk 2 above)
Comment (b1n-ch1kn) [reply src/control/path_follower/path_follower/node_particle_pursuit_inherit.py]: yeah i was. worked it out tho

File: src/common/driverless_common/driverless_common/shutdown_node.py
[the remark this answers is not in the document]
Comment (b1n-ch1kn) [reply src/common/driverless_common/driverless_common/shutdown_node.py:1]: found it
image

**Comment by b1n-ch1kn** [comment]
Had a thought today
What if we set these up as a lifecycle and a 'standard' node pairing. that way the standard can start running when you want to debug at certain points - not managed by the AS state either.
The example I was envisioning was: 
-I'm testing pure pursuit, simulating us completing the discovery lap, but I'm not running my trackdrive handler cause i'm just testing stuff.
This way we can run all nodes in a managed way if we are simulating a race or actually running end-to-end

File: src/common/driverless_common/driverless_common/shutdown_node.py
[the remark this answers is not in the document]
Comment (supex0fan) [reply src/common/driverless_common/driverless_common/shutdown_node.py:1]: Yea... sorry about that.

**Comment by b1n-ch1kn** [comment]
Naming suggestion for mid-spline planner, differentiate between node and lifecycle as prefix (like pursuit)
image

**Review by b1n-ch1kn — APPROVED** [review]
Tested, works as intended

**Pull request merged**